coderd/database/dbfake/dbfake.go

```diff
@@ -39,44 +41,83 @@ func Workspace(t testing.TB, db database.Store, seed database.Workspace) databas
 
 // WorkspaceWithAgent is a helper that generates a workspace with a single resource
 // that has an agent attached to it. The agent token is returned.
-func WorkspaceWithAgent(t testing.TB, db database.Store, seed database.Workspace) (database.Workspace, string) {
+func WorkspaceWithAgent(
+	t testing.TB, db database.Store, seed database.Workspace,
+	mutations ...func([]*sdkproto.Agent) []*sdkproto.Agent,
+) (
+	database.Workspace, string,
+) {
 	t.Helper()
 	authToken := uuid.NewString()
+	agents := []*sdkproto.Agent{{
+		Id: uuid.NewString(),
+		Auth: &sdkproto.Agent_Token{
+			Token: authToken,
+		},
+	}}
+	for _, m := range mutations {
+		agents = m(agents)
+	}
 	ws := Workspace(t, db, seed)
 	WorkspaceBuild(t, db, ws, database.WorkspaceBuild{}, &sdkproto.Resource{
-		Name: "example",
-		Type: "aws_instance",
-		Agents: []*sdkproto.Agent{{
-			Id: uuid.NewString(),
-			Auth: &sdkproto.Agent_Token{
-				Token: authToken,
-			},
-		}},
+		Name:   "example",
+		Type:   "aws_instance",
+		Agents: agents,
 	})
 	return ws, authToken
 }
 
-// WorkspaceBuild inserts a build and a successful job into the database.
-func WorkspaceBuild(t testing.TB, db database.Store, ws database.Workspace, seed database.WorkspaceBuild, resources ...*sdkproto.Resource) database.WorkspaceBuild {
-	t.Helper()
+type BuildBuilder struct {
+	t         testing.TB
+	db        database.Store
+	ps        pubsub.Pubsub
+	ws        database.Workspace
+	seed      database.WorkspaceBuild
+	resources []*sdkproto.Resource
+}
+
+func WorkspaceBuildBuilder(t testing.TB, db database.Store, ws database.Workspace) BuildBuilder {
+	return BuildBuilder{t: t, db: db, ws: ws}
+}
+
+func (b BuildBuilder) Pubsub(ps pubsub.Pubsub) BuildBuilder {
+	//nolint: revive // returns modified struct
+	b.ps = ps
+	return b
+}
+
+func (b BuildBuilder) Seed(seed database.WorkspaceBuild) BuildBuilder {
+	//nolint: revive // returns modified struct
+	b.seed = seed
+	return b
+}
+
+func (b BuildBuilder) Resource(resource *sdkproto.Resource) BuildBuilder {
+	//nolint: revive // returns modified struct
+	b.resources = append(b.resources, resource)
+	return b
+}
+
+func (b BuildBuilder) Do() database.WorkspaceBuild {
+	b.t.Helper()
 	jobID := uuid.New()
-	seed.ID = uuid.New()
-	seed.JobID = jobID
-	seed.WorkspaceID = ws.ID
+	b.seed.ID = uuid.New()
```

Comment:
Is `seed` required or optional?

Reply:
it's optional.  Much of the code that this replaces just passes a zero value for the seed, which is equivalent to not calling `Seed()`